pkg/deploy/deployer.go

```diff
@@ -0,0 +1,311 @@
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	azcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	azresources "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/Azure/ARO-RP/pkg/deploy/generator"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/resources"
+)
+
+var _ Deployer = (*deployer)(nil)
+
+type Deployer interface {
+	PreDeploy(context.Context, *logrus.Entry) (string, error)
+	Deploy(context.Context, *logrus.Entry, string) error
+	Upgrade(context.Context, *logrus.Entry) error
+}
+
+type deployer struct {
+	log         *logrus.Entry
+	deployments resources.DeploymentsClient
+	vmss        compute.VirtualMachineScaleSetsClient
+	vmssvms     compute.VirtualMachineScaleSetVMsClient
+	network     network.PublicIPAddressesClient
+
+	version           string
+	resourceGroupName string
+	subscriptionID    string
+}
+
+func NewDeployer(ctx context.Context, log *logrus.Entry, gitCommit string) (*deployer, error) {
+	for _, key := range []string{
+		"AZURE_TENANT_ID",
+		"AZURE_CLIENT_SECRET",
+		"AZURE_CLIENT_ID",
+		"AZURE_SUBSCRIPTION_ID",
+		"RP_PARAMETERS_FILE",
+		"RP_RESOURCEGROUP_NAME",
+	} {
+		if _, found := os.LookupEnv(key); !found {
+			return nil, fmt.Errorf("environment variable %q unset", key)
+		}
+	}
+
+	gitCommit = "02a47a41"
+
+	if gitCommit == "unknown" ||
+		strings.Contains(gitCommit, "dirty") {
+		return nil, fmt.Errorf("gitCommit '%s' is not valid deployment version", gitCommit)
+	}
+
+	authorizer, err := auth.NewClientCredentialsConfig(os.Getenv("AZURE_CLIENT_ID"), os.Getenv("AZURE_CLIENT_SECRET"), os.Getenv("AZURE_TENANT_ID")).Authorizer()
+	if err != nil {
+		return nil, err
+	}
+
+	d := &deployer{
+		log:               log,
+		version:           gitCommit,
+		resourceGroupName: os.Getenv("RP_RESOURCEGROUP_NAME"),
+		subscriptionID:    os.Getenv("AZURE_SUBSCRIPTION_ID"),
+	}
+
+	d.deployments = resources.NewDeploymentsClient(d.subscriptionID, authorizer)
+	d.vmss = compute.NewVirtualMachineScaleSetsClient(d.subscriptionID, authorizer)
+	d.vmssvms = compute.NewVirtualMachineScaleSetVMsClient(d.subscriptionID, authorizer)
+	d.network = network.NewPublicIPAddressesClient(d.subscriptionID, authorizer)
+
+	return d, nil
+}
+
+// PreDeploy deploys NSG and ManagedIdentity, needed for man deployment
+func (d *deployer) PreDeploy(ctx context.Context, log *logrus.Entry) (rpServicePrincipalID string, err error) {
+	// if RP_PREDEPLOY is not set, we read old rpServicePrincipalID and return it
+	if len(os.Getenv("RP_PREDEPLOY")) != 0 {
```

Comment:
test for the existence of the deployment instead of requiring the env var

Reply:
deployment might exist, but we want to redeploy because we are changing something. So we would set this variable once in CI and remove later.